forward REQUEST messages to primary when prepare timer expires

nhoriguchi opened the pull request:
This pull request proposes to introduce request forwarding to handle some faulty case in request submission. This allows the MinBFT cluster to accept a request in some cases when the primary does not receive it from the client. The table below provides brief summary of how the behavior changes:

| case           | origin | patched | 
|----------------|--------|---------| 
| R0R1R2         | pass   | pass    | 
| R0             | fail   | fail    | 
| R0R1           | pass   | pass    | 
| R1             | fail   | fail    | 
| R1R2           | fail   | pass    | 
| R0R1Q2         | pass   | pass    | 
| R0Q1           | pass   | pass    | 
| R1Q2           | fail   | pass    | 

Here we consider a minimum typical cluster (N is 3 and replica 0 is the primary.) Case ID shows which replica(s) a client sends REQUEST to. For example, in case "R0R1" a client sends REQUEST only to replica 0 and 1. In another example "R0R1Q2", a client sends proper REQUEST messages to replica 0 and 1, and a modified REQUEST message (with the same seq) to replica 2. In the result columns, "pass" means that the cluster reaches a consensus.

The most interesting case is "R1R2", where a client sends the proper REQUEST messages to replica 1 and 2. This case becomes "pass" due to request forwarding.

What I noticed while working on this is that case "R1" fails even with this patch. This is because in current implementation the replica could send REPLY to the client only when the replica received the REQUEST directly from the client (i.e. a replica who got a forwarded request can't send REPLY.) Similarly, a backup replica which didn't receive REQUEST from the client and processed REQUEST embedded in the PREPARE message can't send REPLY message to its requester. I guess that these are not due to some specific reason, but it's just implemented now as such. So we can do some to fix them in the future if we think it's worth doing.

Resolves #140
Resolves #5

ynamiki commented on core/message-handling.go:
```diff
@@ -51,6 +51,10 @@ type incomingMessageHandler func(msg messages.Message, own bool) (reply <-chan m
 
+// forwardedRequestSupplier supplies Request messages to forward to
+// the primary replica.
+type forwardedRequestSupplier func(out chan<- []byte, forwardIn chan messages.Request)
```
Can we change type of `forwardIn` to `<-chan` to increase the type-safety?

nhoriguchi replied:
Sure, I fixed. Thanks!

ynamiki commented on core/message-handling.go:
```diff
@@ -251,28 +255,56 @@ func makeMessageStreamHandler(handle incomingMessageHandler, logger *logging.Log
 			continue
 		}
 
+		forwardOut[peerID] = make(chan messages.Request)
```
Now we can also use `<-chan` here, Line 258, and 264.

nhoriguchi replied:
When I use `<-chan` on L258/L264/L271, compiler shows the following error on L302:

> invalid operation: cannot send to receive-only type <-chan messages.Request

https://github.com/Naoya-Horiguchi/minbft/blob/fe982365837a46d6326ac363937778ab10e70e8c/core/message-handling.go#L300-L303

L307 can have `<-` as you suggested, so I added it.

ynamiki replied:
> When I use <-chan on L258/L264/L271, compiler shows the following error on L302

Sorry, I overlooked the send operation in L302; They must be bidirectional.

ynamiki commented on core/message-handling.go:
```diff
@@ -251,28 +255,56 @@ func makeMessageStreamHandler(handle incomingMessageHandler, logger *logging.Log
 // startPeerConnection initiates asynchronous message exchange with a
 // peer replica.
-func startPeerConnection(connect peerConnector, supply peerMessageSupplier) error {
+func startPeerConnection(connect peerConnector, supply peerMessageSupplier, forward forwardedRequestSupplier, forwardIn chan messages.Request) error {
```
```suggestion
func startPeerConnection(connect peerConnector, supply peerMessageSupplier, forward forwardedRequestSupplier, forwardIn <-chan messages.Request) error {
```

ynamiki reviewed (commented):
Following the last update, there can be some more updates on the type-safety.

ynamiki reviewed (approved):
Thanks, looks good to me.

sergefdrv commented on core/message-handling.go:
```diff
@@ -251,28 +255,56 @@ func makeMessageStreamHandler(handle incomingMessageHandler, logger *logging.Log
 	}
 }
 
+var forwardOut map[uint32](chan messages.Request)
```
This should not be a global variable since it represents a piece of individual replica state.

nhoriguchi replied:
OK, so I'm thinking of adding a new package under core/internal to manage request forwarding, which seems to me better than adding into existing packages.

sergefdrv commented on core/message-handling.go:
```diff
@@ -251,28 +255,56 @@ func makeMessageStreamHandler(handle incomingMessageHandler, logger *logging.Log
+				panic(err)
+			}
+			out <- msgBytes
+		}
```
This loop is almost the same as in `makePeerMessageSupplier`. Maybe we can move the common part to a separate helper function? Or pass the forward channel to `makePeerMessageSupplier` and extend it like follows:

```go
for {
	var msg messages.Message

	select {
	case msg = <-forwardChan:
	case msg = <-logChan:
	}

	msgBytes, err := msg.MarshalBinary()
	if err != nil {
		panic(err)
	}
	out <- msgBytes
}
```

nhoriguchi replied:
Yes, we can factorize this code. What was beyond my first choice was that REQUEST does not belong to message category "Peer Message", so if we go in this direction, maybe I'll rename `makePeerMessageSupplier` to simply `makeMessageSupplier`.

sergefdrv commented on core/message-handling.go:
```diff
@@ -251,28 +255,56 @@ func makeMessageStreamHandler(handle incomingMessageHandler, logger *logging.Log
+// prepare timer expires.
+func forwardRequestMessage(peerID uint32, msg messages.Request, logger *logging.Logger) {
+	logger.Debugf("Forward REQUEST to replica %d\n", peerID)
+	forwardOut[peerID] <- msg
```
I'm afraid this may block the caller for indefinite time if the destination replica is faulty. How do we handle this situation?

nhoriguchi replied:
I didn't consider such situation. I think that this code does not block as long as `forwardOut` channel is listened by the goroutine running forwardedRequestSupplier, which is satisfied if the replica is not faulty. How does the destination replica affect this code?

sergefdrv commented:
> What I noticed while working on this is that case "R1" fails even with this patch. This is because in current implementation the replica could send REPLY to the client only when the replica received the REQUEST directly from the client (i.e. a replica who got a forwarded request can't send REPLY.) Similarly, a backup replica which didn't receive REQUEST from the client and processed REQUEST embedded in the PREPARE message can't send REPLY message to its requester. I guess that these are not due to some specific reason, but it's just implemented now as such. So we can do some to fix them in the future if we think it's worth doing.

This is because replicas use received Request messages to distinguish incoming client connection originated at different clients. We could consider improving this at some point.

nhoriguchi committed:
messagelog: change channel type to messages.Message

We will introduce another user of messagelog that sends messages
rather than ReplicaMssage, so this patch suggests to change channel
type to more general one.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
core: forward REQUEST messages to primary when prepare timer expires

This patch introduces request forwarding functionality for replica cluster
to accept a request even if the primary replica occasionally failed to receive
it from the client.  As a result, the MinBFT cluster gets more tolerant for
client's faulty behavior.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi commented:
Change summary:

- generalized peer message supplier to handle both log channel and forward channel,
- use `messageLog` instances to handle unicast, which indirectly solves the concern about blocking channel operations because `supplyMessages()` already addresses it.

sergefdrv commented:
@Naoya-Horiguchi Thank you for doing this, but I'm afraid I barely have time to review it until the beginning of April. On the other hand, I'd still like to vet on this issue and give my feedback. Would it be fine for you to wait for my feedback for a couple more weeks?

nhoriguchi commented:
> @Naoya-Horiguchi Thank you for doing this, but I'm afraid I barely have time to review it until the beginning of April. On the other hand, I'd still like to vet on this issue and give my feedback. Would it be fine for you to wait for my feedback for a couple more weeks?

Sure, this issue is not urgent, so we can wait until April.

sergefdrv commented on core/message-handling.go:
```diff
@@ -45,11 +45,12 @@ type messageStreamHandler func(in <-chan []byte, reply chan<- []byte)
 // delivered to the peer replica.
-type peerMessageSupplier func(out chan<- []byte)
+// Another parameter 'forwardChan' is to accept unicasted messages.
+type messageSupplier func(out chan<- []byte, forwardChan <-chan messages.Message)
```
I suggest passing `forwardChan` to `makeMessageSupplier` instead. This way would better align with what we already do with the common message log.

sergefdrv reviewed (commented):
I left a minor comment, but apart from that, I think this PR is a solid step towards fully solving the request forwarding problem.

Now, using `messagelog` for buffering the forwarded requests eliminates dandling goroutines when the destination replica stops receiving the messages. However, this will still accumulate the forwarded messages in the unicast message logs, whereas we only need to keep at most one request per client (until we add support for concurrent client requests). I think we can try to improve this as a separate issue.

nhoriguchi commented:
I've updated the branch.

> Now, using messagelog for buffering the forwarded requests eliminates dandling goroutines when the destination replica stops receiving the messages. However, this will still accumulate the forwarded messages in the unicast message logs, whereas we only need to keep at most one request per client (until we add support for concurrent client requests). I think we can try to improve this as a separate issue.

I agree, I added an inline comment to mention this.

> I suggest passing forwardChan to makeMessageSupplier instead. This way would better align with what we already do with the common message log.

Applied the suggestion, thank you.

sergefdrv commented on core/message-handling.go:
```diff
@@ -253,14 +253,20 @@ func makeMessageStreamHandler(handle incomingMessageHandler, logger *logging.Log
+		// make the requests accumulated when the destination replica stops
+		// receiving them. Currently that does not happen because we only need
+		// to keep at most one request per client. We will need some update when
+		// we add support for concurrent client requests.
```
Sorry for confusion, but indefinitely many Request messages can accumulate there, even though we currently don't support concurrent requests from the same client. This is because each _new_ Request may be added to the unicast message log for forwarding but never removed from there (check `core/internal/clientstate/request-seq.go`). 

I was thinking of a new component to replace the unicast message logs here. The new component would act as an asynchronous buffer, limited to one Request message per client. Maybe there are other possible solutions. Maybe we should update this comment, merge the PR and tackle the remaining problem as a separate issue.

nhoriguchi replied:
Thanks for the explanation, I updated the comment (pushed to the branch):

https://github.com/Naoya-Horiguchi/minbft/blob/683019e6dd07b1bb6d05ccb55b932fb17bbb24cc/core/message-handling.go#L262-L265

> I was thinking of a new component to replace the unicast message logs here. The new component would act as an asynchronous buffer, limited to one Request message per client. Maybe there are other possible solutions. 

I don't have any specific idea for now, so I'll look into "new component" approach first.
